llvm/lib/Transforms/Utils/LoopIterCountsToAssumptions.cpp

```diff
@@ -0,0 +1,215 @@
+//===-- LoopIterCountsToAssumptions.cpp - add Loop assumes -*- C++ -*-===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// (c) Copyright 2024 Advanced Micro Devices, Inc. or its affiliates
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/Transforms/Utils/LoopIterCountsToAssumptions.h"
+#include "llvm/Analysis/AliasAnalysis.h"
+#include "llvm/Analysis/AssumptionCache.h"
+#include "llvm/Analysis/MemorySSA.h"
+#include "llvm/IR/Dominators.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/Value.h"
+#include "llvm/Support/Casting.h"
+#include "llvm/Transforms/Utils/LoopUtils.h"
+#include "llvm/Transforms/Utils/ScalarEvolutionExpander.h"
+
+#define DEBUG_TYPE "loop-metadata"
+
+using namespace llvm;
+namespace {
+/// check basic loop rotation conditions
+bool isLoopRotateCandidate(const Loop &L) {
+
+  SmallVector<BasicBlock *> LoopLatches;
+  L.getLoopLatches(LoopLatches);
+  SmallVector<BasicBlock *> ExitBlocks;
+  L.getExitBlocks(ExitBlocks);
+  BranchInst *BI = dyn_cast<BranchInst>(L.getHeader()->getTerminator());
+  return L.isLoopExiting(L.getHeader()) && (BI && BI->isConditional()) &&
```

Comment:
As we can process to many unfeasible loops, I suggest to use early returns. If one condition is not valid, we don't need to test the others.

Reply:
I converted it into many early returns